GCEWindowsAgent/serial_listener.go

```diff
@@ -0,0 +1,462 @@
+	if seenPreSnapshotoperationIds.Contains(req.OperationId) {
+		logger.Infof("ignoring pre-snapshot request with operation id %d", req.OperationId)
+		return
+	}
+
+	seenPreSnapshotoperationIds.Add(req.OperationId, nil)
+
+	logger.Infof("running pre-snapshot script")
+	config := getConfig()
+	code := runScript(config.PreSnapshotScriptUrl, config.Timeout)
+	res := serialResponse{identifier, "PRERESP", req.Version, code, req.OperationId}
+	out, err := json.Marshal(res)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	writeSerial(serialPort, out)
+}
+
+func handleResumePostSnapshotRequest(req serialRequest) {
+	if seenPostSnapshotoperationIds.Contains(req.OperationId) {
+		logger.Infof("ignoring post-snapshot request with operation id %d", req.OperationId)
+		return
+	}
+	seenPostSnapshotoperationIds.Add(req.OperationId, nil)
+
+	logger.Infof("running post-snapshot script")
+	config := getConfig()
+	code := runScript(config.PostSnapshotScriptUrl, config.Timeout)
+	res := serialResponse{identifier, "POSTRESP", req.Version, code, req.OperationId}
+	out, err := json.Marshal(res)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	writeSerial(serialPort, out)
+}
+
+func handleMessage(msg serialRequest) {
+	switch sig := msg.Signature; sig {
+	case "PREREQ":
+		handlePrepareForSnapshotRequest(msg)
+	case "POSTREQ":
+		handleResumePostSnapshotRequest(msg)
+	default:
+		logger.Fatalf("unknown message signature %s", msg.Signature)
+	}
+}
+
+func parseSerialData(serialData []byte) {
+	for _, c := range serialData {
+		if c == '\n' {
+			// Should be the end of the message
+			handleMessage(*parseMessage())
+		} else if c != '\x00' {
+			currentMsg = append(currentMsg, c)
+		}
+	}
+}
+
+func getMetadataAttribute(key string) ([]byte, error) {
+	client := &http.Client{
+		Timeout: defaultTimeout,
+	}
+
+	url := metadataAttributesURL + key
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Metadata-Flavor", "Google")
+
+	var res *http.Response
+	// Retry forever.
+	var rt time.Duration
+	for i := 1; ; i++ {
+		res, err = client.Do(req)
+		if err == nil {
+			break
+		}
+		if i < 6 {
+			rt = time.Duration(3*i) * time.Second
+		}
+		logger.Errorf("error connecting to metadata server, retrying in %s, error: %v", rt, err)
+		time.Sleep(rt)
+	}
+	defer res.Body.Close()
+
+	md, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	return md, nil
+}
+
+func createTestConfig() snapshotConfig {
+	var config snapshotConfig
+	config.Timeout = 5
+	config.ContinueOnScriptError = true
+	config.PreSnapshotScriptUrl = testPreSnapshotScript
+	config.PostSnapshotScriptUrl = testPostSnapshotScript
+	return config
+
+}
+
+func createDefaultConfig() snapshotConfig {
+	var config snapshotConfig
+	config.Timeout = 60
+	config.ContinueOnScriptError = true
+	config.PreSnapshotScriptUrl = ""
+	config.PostSnapshotScriptUrl = ""
+	return config
+}
+
+func getConfig() snapshotConfig {
```

Comment:
names of functions are too generic if in main package

Reply:
Renamed!